core/execution.go

```diff
@@ -169,7 +169,12 @@ func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPric
 	// when we're in homestead this also counts for code storage gas errors.
 	if err != nil && (env.ChainConfig().IsHomestead(env.BlockNumber()) || err != vm.CodeStoreOutOfGasError) {
 		contract.UseGas(contract.Gas)
+		env.RevertToSnapshot(snapshotPreTransfer)
 
+		// Nothing should be returned when an error is thrown.
+		return nil, addr, err
+	} else if env.ChainConfig().IsEIP170(env.BlockNumber()) && len(ret) > params.MaxCodeSize {
```

Comment:
A few lines up, back at L:157, there's `if err == nil`, where the clause which does `setCode` is performed. I suggest to check `MaxCodeSize` up there instead of down here, and then also skip the `SetCode` stuff.